File: Java/Problem_093.java
import java.util.*;

public class Euler93 {
    static class Frac {
        long n, d; // reduced, d>0
        Frac(long n, long d) {
            if (d == 0) throw new ArithmeticException("div0");
            if (d < 0) { n = -n; d = -d; }
            long g = gcd(Math.abs(n), d);
            this.n = n / g;
            this.d = d / g;
        }
        static long gcd(long a, long b) {
            while (b != 0) { long t = a % b; a = b; b = t; }
            return a;
        }
        @Override public int hashCode() {
            return Objects.hash(n, d);
        }
        @Override public boolean equals(Object o) {
            if (!(o instanceof Frac)) return false;
            Frac f = (Frac)o;
            return n == f.n && d == f.d;
        }
    }

    static Frac add(Frac a, Frac b){ return new Frac(a.n*b.d + b.n*a.d, a.d*b.d); }
    static Frac sub(Frac a, Frac b){ return new Frac(a.n*b.d - b.n*a.d, a.d*b.d); }
    static Frac mul(Frac a, Frac b){ return new Frac(a.n*b.n, a.d*b.d); }
    static Frac div(Frac a, Frac b){
        if (b.n == 0) throw new ArithmeticException("div0");
        return new Frac(a.n*b.d, a.d*b.n);
    }

    static HashSet<Frac> allValues(ArrayList<Frac> nums) {
        if (nums.size() == 1) {
            HashSet<Frac> base = new HashSet<>();
            base.add(nums.get(0));
            return base;
        }

        HashSet<Frac> out = new HashSet<>();
        int L = nums.size();

        for (int i=0;i<L;i++){
            for (int j=i+1;j<L;j++){
                Frac a = nums.get(i), b = nums.get(j);

                ArrayList<Frac> rest = new ArrayList<>(L-1);
                for (int k=0;k<L;k++) if (k!=i && k!=j) rest.add(nums.get(k));

                java.util.function.Consumer<Frac> merge = (Frac v) -> {
                    ArrayList<Frac> next = new ArrayList<>(rest);
                    next.add(v);
                    out.addAll(allValues(next));
                };

                merge.accept(add(a,b));
                merge.accept(mul(a,b));
                merge.accept(sub(a,b));
                merge.accept(sub(b,a));
                if (b.n != 0) merge.accept(div(a,b));
                if (a.n != 0) merge.accept(div(b,a));
            }
        }
        return out;
    }

    public static void main(String[] args) {
        int bestLen = 0;
        int ba=0, bb=0, bc=0, bd=0;

        for (int a=1;a<=6;a++){
            for (int b=a+1;b<=7;b++){
                for (int c=b+1;c<=8;c++){
                    for (int d=c+1;d<=9;d++){
                        ArrayList<Frac> start = new ArrayList<>();
                        start.add(new Frac(a,1));
                        start.add(new Frac(b,1));
                        start.add(new Frac(c,1));
                        start.add(new Frac(d,1));

                        HashSet<Frac> vals = allValues(start);

                        HashSet<Integer> ints = new HashSet<>();
                        for (Frac v : vals) {
                            if (v.d == 1 && v.n > 0 && v.n <= Integer.MAX_VALUE) {
                                ints.add((int)v.n);
                            }
                        }

                        int n = 1;
                        while (ints.contains(n)) n++;
                        int runLen = n - 1;

                        if (runLen > bestLen) {
                            bestLen = runLen;
                            ba=a; bb=b; bc=c; bd=d;
                        }
                    }
                }
            }
        }

        System.out.println("" + ba + bb + bc + bd); // 1258
    }
}

File: Python/Problem_093.py
from math import gcd
from itertools import combinations

class Frac:
    __slots__ = ("n", "d")
    def __init__(self, n: int, d: int = 1):
        if d == 0:
            raise ZeroDivisionError
        if d < 0:
            n, d = -n, -d
        g = gcd(abs(n), d)
        self.n = n // g
        self.d = d // g

    def __add__(self, other): return Frac(self.n * other.d + other.n * self.d, self.d * other.d)
    def __sub__(self, other): return Frac(self.n * other.d - other.n * self.d, self.d * other.d)
    def __mul__(self, other): return Frac(self.n * other.n, self.d * other.d)
    def __truediv__(self, other):
        if other.n == 0:
            raise ZeroDivisionError
        return Frac(self.n * other.d, self.d * other.n)

    def __hash__(self): return hash((self.n, self.d))
    def __eq__(self, other): return self.n == other.n and self.d == other.d

def all_values(nums):
    # nums: list[Frac]
    if len(nums) == 1:
        return {nums[0]}

    out = set()
    L = len(nums)
    for i in range(L):
        for j in range(i + 1, L):
            a, b = nums[i], nums[j]
            rest = [nums[k] for k in range(L) if k != i and k != j]

            # commutative ops once
            out |= all_values(rest + [a + b])
            out |= all_values(rest + [a * b])

            # non-commutative both orders
            out |= all_values(rest + [a - b])
            out |= all_values(rest + [b - a])

            if b.n != 0:
                out |= all_values(rest + [a / b])
            if a.n != 0:
                out |= all_values(rest + [b / a])

    return out

def euler_93():
    best_len = 0
    best = None

    for digits in combinations(range(1, 10), 4):
        fracs = [Frac(x) for x in digits]
        vals = all_values(fracs)

        ints = set()
        for v in vals:
            if v.d == 1 and v.n > 0:
                ints.add(v.n)

        n = 1
        while n in ints:
            n += 1
        run_len = n - 1

        if run_len > best_len:
            best_len = run_len
            best = digits

    return "".join(map(str, best))

if __name__ == "__main__":
    print(euler_93())  # 1258
